infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/UniqueNamesProvisioner.java

```diff
@@ -48,15 +59,38 @@ public void provision(T k8sEnv, RuntimeIdentity identity) throws InfrastructureE
 
     TracingTags.WORKSPACE_ID.set(workspaceId);
 
-    final Set<Pod> pods = new HashSet<>(k8sEnv.getPods().values());
-    k8sEnv.getPods().clear();
-    for (Pod pod : pods) {
+    final Map<String, ConfigMap> configMaps = k8sEnv.getConfigMaps();
+    Map<String, String> configMapNameTranslation = new HashMap<>();
+    for (ConfigMap configMap : configMaps.values()) {
+      final String originalName = configMap.getMetadata().getName();
+      putLabel(configMap.getMetadata(), Constants.CHE_ORIGINAL_NAME_LABEL, originalName);
+      final String uniqueName = Names.uniqueResourceName(originalName, workspaceId);
+      configMap.getMetadata().setName(uniqueName);
+      configMapNameTranslation.put(originalName, uniqueName);
+    }
+
+    final Collection<PodData> podData = k8sEnv.getPodsData().values();
+    for (PodData pod : podData) {
       final ObjectMeta podMeta = pod.getMetadata();
-      putLabel(pod, Constants.CHE_ORIGINAL_NAME_LABEL, podMeta.getName());
-      final String podName = Names.uniquePodName(podMeta.getName(), workspaceId);
+      putLabel(podMeta, Constants.CHE_ORIGINAL_NAME_LABEL, podMeta.getName());
+      final String podName = Names.uniqueResourceName(podMeta.getName(), workspaceId);
       podMeta.setName(podName);
-      k8sEnv.getPods().put(podName, pod);
+      if (configMapNameTranslation.size() > 0) {
+        rewriteConfigMapNames(pod, configMapNameTranslation);
+      }
     }
+
+    // We expliticly need to modify the deployments in the environment to provision unique names
+    // for them.
+    final Collection<Deployment> deployments = k8sEnv.getDeploymentsCopy().values();
```

Comment:
I think we need tests to cover all the changes in this class

Reply:
Good point, will do.

Reply:
Add a bunch of tests for configmap/deployment renaming, and also for renaming configmaps referenced in pods.